lib/ZnapZend.pm

```diff
@@ -474,11 +479,24 @@ my $sendRecvCleanup = sub {
             # operations - so one destroy operation takes ages...
             # but hundreds of queued operations take the same time
             # and are all committed at once.
+            $self->zLog->debug('checking to clean up snapshots recursively from source '. $backupSet->{src});
 
             @snapshots = @{$self->zZfs->listSnapshots($backupSet->{src}, $backupSet->{snapCleanFilter})};
             $toDestroy = $self->zTime->getSnapshotsToDestroy(\@snapshots,
                          $backupSet->{srcPlanHash}, $backupSet->{tsformat}, $timeStamp);
 
+            # Save the names we have seen, to not revisit them below for children
+            for (@snapshots) {
+                my $snapname = $_;
```

Comment:
is this to make sure that the content of the snapshots array does not get changed ?

Reply:
yes, I found that with `for my $snapname (@toDestroy)` at least, regexing the `$snapname` in the next line did change the array entries from `dataset@snapname` to `snapname`.

Reply:
this is a feature :) see my proposed function above